rules/prefer-error-cause.js

```diff
@@ -0,0 +1,392 @@
+		// It could be VariableDeclarator or AssignmentExpression.
+		const newExpressionSelector = reference =>
+			(reference.identifier?.parent.type === 'AssignmentExpression'
+			|| reference.identifier?.parent.type === 'VariableDeclarator')
+			&& reference.writeExpr?.type === 'NewExpression'
+			&& reference.resolved?.name === throwStatement.argument.name;
+
+		const thrownErrorDeclarators = getReferences(context.getScope())
+			.filter(reference => newExpressionSelector(reference))
+			.map(reference => reference.identifier.parent);
+
+		// Report newExpression not found in given block.
+		// `try {} catch { const err = new Error; try {} catch { throw err; } }`
+		if (_.isEmpty(thrownErrorDeclarators)) {
+			reportCannotBeFixed(catchBlock, context);
+			return [];
+		}
+
+		// `try {} catch (err) { let e1 = new Error('oops'); e1 = new Error('oops'); throw error;}`
+		for (const thrownErrorDeclarator of thrownErrorDeclarators) {
+			statementToFix = thrownErrorDeclarator;
+
+			let targetArguments;
+			if (thrownErrorDeclarator.type === 'VariableDeclarator') {
+				targetArguments = thrownErrorDeclarator.init.arguments;
+			} else if (thrownErrorDeclarator.type === 'AssignmentExpression') {
+				targetArguments = thrownErrorDeclarator.right.arguments;
+			}
+
+			errorConstructorLastArgument = targetArguments[targetArguments.length - 1];
+
+			if (!errorConstructorLastArgument) {
+				reportCannotBeFixed(catchBlock, context);
+				return [];
+			}
+
+			result.push({
+				node: errorConstructorLastArgument,
+				statementToFix,
+			});
+		}
+	}
+
+	return result;
+};
+
+const handleCatchBlock = ({context, catchBlock, parameter, throwStatement}) => {
+	const errorArgumentIdentifier = parameter?.name;
+	const throwStatementArgument = throwStatement.argument;
+
+	// Filter rethrowing the error itself.
+	// `try {} catch (err) { throw err; }`
+	// `promise.catch(err => { throw err; })`
+	if (errorArgumentIdentifier && errorArgumentIdentifier === throwStatementArgument.name) {
+		return;
+	}
+
+	// Report none identifier parameter
+	// `try {} catch ({error}) {}`
+	// `promise.catch({error} => {})`
+	if (parameter && parameter.type !== 'Identifier') {
+		context.report({
+			node: catchBlock,
+			messageId: ERROR,
+		});
+		return;
+	}
+
+	const nodesToFixCandidates = getAllNodesToFix({catchBlock, context, throwStatement});
+
+	const nodesToFix = nodesToFixCandidates.filter(({node}) =>
+		!hasCauseProperty(node),
+	);
+
+	for (const nodeToFix of nodesToFix) {
+		context.report({
+			node: catchBlock,
+			messageId: ERROR,
+			fix(fixer) {
+				return fix({
+					fixer,
+					context,
+					catchBlock,
+					statementToFix: nodeToFix.statementToFix,
+					errorConstructorLastArgument: nodeToFix.node,
+					errorArgumentIdentifier,
+				});
+			},
+		});
+	}
+};
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => ({
+	'ThrowStatement'(node) {
+		const catchBlock = getCatchBlock(node);
```

Comment:
Maybe easier if we listen on `NewExpression` instead of `ThrowStatement`? Do we don't need trace variables.

Reply:
IMHO, when listen `NewExpression`, we need to find out if the line is throw statement and if not, find throw statements in the block and check each throw statement's argument.

So I think maybe the code would be complicated as much as now it is.